source/distributed/shared_tria.cc

```diff
@@ -351,6 +351,51 @@ namespace parallel
     void
     Triangulation<dim, spacedim>::execute_coarsening_and_refinement()
     {
+      // make sure that all refinement/coarsening flags are the same on all
+      // processes
+      {
+        std::vector<unsigned int> refinement_configurations(
+          this->n_active_cells() * 2, 0u);
+        for (const auto &cell : this->active_cell_iterators())
+          if (cell->is_locally_owned())
+            {
+              refinement_configurations[cell->active_cell_index() * 2 + 0] =
+                cell->refine_flag_set();
+              refinement_configurations[cell->active_cell_index() * 2 + 1] =
+                cell->coarsen_flag_set();
+            }
+
+        Utilities::MPI::max(refinement_configurations,
+                            this->get_communicator(),
+                            refinement_configurations);
+
+        for (const auto &cell : this->active_cell_iterators())
+          {
+            cell->clear_refine_flag();
+            cell->clear_coarsen_flag();
+
+            Assert(
+              (refinement_configurations[cell->active_cell_index() * 2 + 0] >
+                   0 ?
+                 1 :
+                 0) +
+                  refinement_configurations[cell->active_cell_index() * 2 +
+                                            1] <=
+                1,
+              ExcMessage(
+                "Refinement/coarsening flags of cells are not consistent in parallel!"));
```

Comment:
I don't understand how this tests that the flags are the same everywhere, as the initial comment in this function is suggesting. Isn't all that you check that do not set refinement and coarsen flags simultaneously anywhere? That is probably sufficient, but not what the comment says. (By doing the `allreduce` in place, you lose track of what that process stored originally. You could retrieve the information again from what is stored in the flags, but I'm not sure that comparison is what you want. Either way, I think it would be good to have a way to be able to only store flags on locally owned cells.)

Reply:
Indeed, it only assures that for the same cell not a refinement flag and a coarsening flag are set. The reasoning is that if only flags of locally owned cells are set, this is the right behavior. In the other case, it is a good heuristic under the assumption that if any process sets a flag it is taken (if both refinement and coarsening flag are set, the situation is not clear).